native-image-tests/build.gradle

```diff
@@ -0,0 +1,48 @@
+plugins {
+  id "com.palantir.graal" version "0.7.1"
+}
+
+dependencies {
+  implementation deps.assertj
+  implementation "org.junit.jupiter:junit-jupiter-api:5.7.0"
+  implementation "org.junit.jupiter:junit-jupiter-engine:5.7.0"
+  implementation "org.junit.platform:junit-platform-console:1.7.0"
+
+  compileOnly deps.jsr305
+}
+
+animalsniffer {
+  ignoreFailures = true
+}
+
+def graalHome = "/Library/Java/JavaVirtualMachines/graalvm-ce-java11-20.2.0/Contents/Home"
+def graalvmResources = project.file("$buildDir/graalvm/resources")
+def graalvmConfig = project.file("$graalvmResources/META-INF/native-image")
+graalvmConfig.mkdirs()
+
+sourceSets {
+  main.resources.srcDirs += graalvmResources
+}
+
+task seedGraalvmConfig(type: Exec) {
+  dependsOn("classes")
+  commandLine(
+    "$graalHome/bin/java",
+    "-agentlib:native-image-agent=config-output-dir=$graalvmConfig",
+    "-classpath", sourceSets.main.runtimeClasspath.getAsPath(),
+    "org.junit.platform.console.ConsoleLauncher",
```

Comment:
Instead of this approach of running twice, the technique |I've used is

https://github.com/yschimke/okurl/blob/master/src/main/kotlin/com/baulsupp/okurl/graal/RuntimeReflectionRegistrationFeature.kt

```
@AutomaticFeature
internal class RuntimeReflectionRegistrationFeature : Feature {
  override fun beforeAnalysis(access: BeforeAnalysisAccess) {
    try {
      val pkg = "com.baulsupp.okurl"
      ClassGraph()
//      .verbose() // Log to stderr
        .enableAnnotationInfo()
        .enableMethodInfo()
        .enableClassInfo() // Scan classes, methods, fields, annotations
        .acceptPackages(pkg) // Scan com.xyz and subpackages (omit to scan all packages)
        .scan()
        .use { scanResult ->                    // Start the scan
          for (classInfo in scanResult.getSubclasses(JsonAdapter::class.java.name)) {
            registerMoshiAdapter(classInfo.loadClass())
          }
          for (classInfo in scanResult.getClassesWithMethodAnnotation(FromJson::javaClass.name)) {
            registerAnnotatedMoshiAdapter(classInfo.loadClass())
          }
        }
```


So you could scan under some package, initially okhttp3.sample for @Test annotations, and widen over time.

Reply:
I'm not against this however, just pointing out options.

Reply:
Digging into this a bit more, I understand the issues with the scan the classpath for annotations approach.  It makes deep assumptions based on the typical uses of Junit.  But Junit 5 has moved past a lot of this, has a clean API for test discovery etc https://junit.org/junit5/docs/5.7.0-RC1/api/org.junit.platform.launcher/org/junit/platform/launcher/core/EngineDiscoveryOrchestrator.html

So maybe the final result of a clean plugin will be a combination of the two.  an @AutomaticFeature that uses EngineDiscoveryOrchestrator instead of just launching the tests, to register each class for reflection needed for later test execution?